otel worker post (#499)
fiberplane/fiberplane: www/src/content/blog/2025-02-18-otel-cloudflare-worker.mdx

Reviewer: not sure if you have any changes locally so didn't wanna do this myself, but i'd suggest having claude do a pass to find typos and confusing sentences

tags:
  - Cloudflare
  - OpenTelemetry
  - Rust
---

import PackageManagers from "@/components/PackageManagers.astro";
import { LinkCard } from "@astrojs/starlight/components";
import playgroundVideo from "../../assets/blog/2025-02-12-playground.mov";

In today's distributed systems landscape, observability is crucial for understanding 
and debugging complex applications. Today, I'm excited to share details about 
[otel-worker](https://github.com/fiberplane/otel-worker). Think of it as an OpenTelemetry collector, 
but without the YAML complexity and heavy infrastructure.

We built otel-worker after encountering some persistent latency with Cloudflare’s D1,
where the real bottleneck was elusive. 
We wanted a straightforward, powerful way to capture, store, and analyze our 
production traces. 

There are existing OpenTelemetry observability solutions including Datadog (if you enjoy throwing money at problems), Baselime, and open source projects, but we decided to roll our own, written in 
Rust and specifically geared towards Cloudflare Workers. This has some key benefits:

- Traditional OpenTelemetry collectors require significant infrastructure

> - Why process teletetry in `us-east-1` when you can can store them directly at the edge with Cloudflare’s D1?

Reviewer: nit: cloudflare has not yet implemented read replicas for D1, so the data is not really at the edge 

also typo: `teletetry`

- No YAML

> We chose Rust for as it's a team-favorite tool inside of Fiberplane and like writing sharp tools

Reviewer: this sentence needs some love

in the language such as our TUI log tailing CLI, [Fogwatch](/blog/fogwatch/).

Alongside Rust, we use the [Axum](https://github.com/tokio-rs/axum) crate for our API. The code is split into 
several crates: `otel-worker` and `otel-worker-core`. The `otel-worker-core` 
crate contains all the shared code that is used by both the `otel-worker` 
crate and our `otel-worker-cli` application. The `otel-worker` crate contains all 
the specifics for running our OTLP ingestion code on Cloudflare. 

> It contains the workers-rs code, a Store implementation that targets 

Reviewer: ```suggestion
This includes: the workers-rs code, a Store implementation that targets 
```

keeps repeating the word contains, could be good to vary it a bit

Cloudflare's D1 database, and a WebSocket implementation that leverages 

> Durable Objects for maintaining persistent connections.

Reviewer: ```suggestion
[Durable Objects](https://developers.cloudflare.com/durable-objects/) for maintaining persistent connections.
```

The CLI lets you list all traces or dive deep into specific spans with simple 
commands like `cargo cli client traces list` or `cargo cli client spans get <trace-id> <span-id>`. 
For cleanup operations, you can delete individual spans or entire traces just as 
easily. The CLI also includes handy system management commands for tasks like database 
cleanup, and for the debugging enthusiasts, there's a hidden WebSocket monitoring tool 
that lets you watch real-time trace updates as they flow through the system. 

### Getting started

The `otel-worker` implements the [OTLP/HTTP spec](https://opentelemetry.io/) 
and supports both JSON and Protocol Buffers encoding (gRPC is not supported). 
It includes HTTP endpoints to retrieve traces and a web-socket endpoint for realtime 
notifications about new traces.

#### Authentication

The `otel-worker` uses simple bearer token authentication, which is *required* 
for both the OTLP/HTTP endpoints and the traces endpoints. 
You'll configure this token using an environment variable during setup.

#### Local Development Setup

To get started with local development, you'll need:

1. Rust installed
2. The Wrangler CLI installed

> Onece you've cloned the [repository](https://github.com/fiberplane/otel-worker),

Reviewer: ```suggestion
Once you've cloned the [repository](https://github.com/fiberplane/otel-worker),
```

```sh
AUTH_TOKEN="your-secret-token-here"
```

> **Note**: Compiling the Worker is not supported on Windows at the moment without WSL.

Finally, start the development server:

```sh
npx wrangler dev
```

Your local server will be running at `http://localhost:8787`. Send it some traces:

The worker accepts traces via OTLP/HTTP. The endpoint to send traces to is
`/v1/traces`.

Here's an example of how to send traces to the worker using `curl` from within
the `examples/send-trace` directory inside the repository:

```sh
curl -X POST http://localhost:8787/v1/traces \
  -H "Authorization: Bearer your-secret-token-here" \
  -H "Content-Type: application/json" \
  --data-binary @trace.json
```

We can now retrieve the traces from the D1 database as follows:

```sh
curl http://localhost:8787/v1/traces \
  -H "Authorization: Bearer your-secret-token-here"
```
![Terminal with otel-worker running](@/assets/blog/2025-02-18-otel-worker.png)

Congratulations! You’ve got a streamlined OpenTelemetry collector up and 
running—no need to wrestle with a mile-long YAML file this time.

## Conclusion

> To wrap things up, we’re well aware that Cloudflare may eventually launch an 

Reviewer: ```suggestion
We’re well aware that Cloudflare is going to launch an 
```
